src/System.Linq/src/System/Linq/Reverse.cs

```diff
@@ -111,6 +111,12 @@ public List<TSource> ToList()
                 return list;
             }
 
+            public HashSet<TSource> ToHashSet(IEqualityComparer<TSource> comparer)
+            {
+                // HashSet is not guaranteed to be ordered, ignoring reverse
+                return new HashSet<TSource>(_source, comparer);
```

Comment:
If we change this to `return _source.ToHashSet(comparer)` then if `_source` is itself a type with an optimised path from this PR it will be able to take that path.